### ISSUE TITLE
Potential bug in baseline: RTC does not outperform BID after bug fix

### ISSUE THREAD

**Issue by wenxichen2746** [Physical-Intelligence/real-time-chunking-kinetix#7]
We identified an implementation issue in RTC related to how previous action chunks are stored. In the original implementation, each chunk was padded with zeros at the end to align the indices of overlapping actions between consecutive chunks. However, this design introduces an error when computing the backward loss in BID.

<img width="1875" height="432" alt="Image" src="https://github.com/user-attachments/assets/4e5b411a-7150-400d-b753-cd02a93149d6" />

The corrected BID implementation in Kinetix is shown below. The corrected version only calculates the norm on the overlapping portion between two action chunks. For reference, the commented section contains the original RTC code. The previous version incorrectly computed the norm between new actions and zero-padded previous actions, which penalized all non-zero sampled actions. As a result, the BID sampler tended to favor actions with smaller magnitudes. 

```
        # def backward_loss(action_chunks: jax.Array):
        #     error = jnp.linalg.norm(action_chunks - prev_action_chunk, axis=-1)  # [n, b, h]
        #     return jnp.sum(error * weights[None, None, :], axis=-1)  # [n, b]
        def backward_loss(action_chunks: jax.Array):
            # Calculate the length of the valid, non-padded portion of the previous action chunk.
            valid_horizon = self.action_chunk_size - execute_horizon
            # Slice the action chunks and weights to only compare the valid, overlapping parts.
            actions_to_compare = action_chunks[:, :, :valid_horizon]
            prev_actions_to_compare = prev_action_chunk[None, :, :valid_horizon] # Add sample dim for broadcasting
            weights_to_apply = weights[None, None, :valid_horizon] 
            # Compute the error only on the sliced, valid portions.
            error = jnp.linalg.norm(actions_to_compare - prev_actions_to_compare, axis=-1)  # [n, b, valid_horizon]
            return jnp.sum(error * weights_to_apply, axis=-1)  # [n, b]
```

This issue may have gone unnoticed in the RTC paper due to the difference in the weight schedules between BID and RTC. BID employs exponentially decaying weights that remain non-zero across the entire chunk, whereas RTC’s weights decay to near-zero in later portions. Consequently, the zero padding has little effect on RTC’s results but biases BID’s backward loss. 

To illustrate the issue, consider a one-dimensional action chunk with four timesteps. The previous action chunk is [1.0, 2.0, 3.0, 4.0], where the action magnitude increases with time. With an execution horizon of 2, the first two actions (1.0 and 2.0) are executed. The RTC implementation then stores the remaining actions as [3.0, 4.0, 0.0, 0.0] by padding zeros to align the indices of overlapping actions in the next chunk.
At the next timestep, suppose the newly sampled action chunk is [3.0, 4.0, 5.0, 6.0]. This sequence aligns well with the previous decision—specifically, the overlapping actions 3.0 and 4.0 are identical. Assuming a discount factor of 0.9, the correct backward loss should be 0.9*(3.0 - 3.0)^2 + 0.9^2* (4.0 - 4.0) = 0
However, in the RTC paper’s implementation, the backward loss in BID is computed between the new action chunk and the zero-padded previous chunk. Consequently, the calculated backward loss becomes 0.9*(3.0 - 3.0)^2 + 0.9^2* (4.0 - 4.0)  + 0.9^3* (5.0 - 0) + 0.9^4* (6.0 - 0) = 7.5816. This example demonstrates how zero padding introduces artificial penalties that bias BID toward smaller-magnitude actions.

Meanwhile, by substituting the parameters H=4 (chunk size), d=1 (inference delay), and s=2 (execution horizon) into Equation (5) of the RTC paper, the resulting guidance weights are [1.0, 0.19, 0.0, 0.0]. Multiplying these weights with the corresponding loss terms effectively nullifies the contributions from the padded elements, leaving the overall loss unchanged.

We report the corrected results below. After fixing the implementation, BID achieves slightly better performance than RTC for execution horizons greater than one. (*Another difference from the original paper is that we introduced a more diverse reward during expert training, as we found that most experts otherwise exhibited nearly identical behaviors. Without multimodal behaviors, employing generative behavior cloning would be unjustified.)

<img width="2011" height="1811" alt="Image" src="https://github.com/user-attachments/assets/1893f091-86a2-4981-a423-7f3bbdf6db01" />

**Comment by kvablack** [comment]
Hi Wenxi, thanks for your detailed writeup.

> BID employs exponentially decaying weights that remain non-zero across the entire chunk, whereas RTC’s weights decay to near-zero in later portions. Consequently, the zero padding has little effect on RTC’s results but biases BID’s backward loss.

I don't think this is true. First, the BID paper does describe computing the backward loss only on overlapping steps (see equation 6). Second, RTC does not decay to "near-zero" in the padded region, but completely to zero in the padded region (see equation 5); i.e., it does not "pay attention" to the padding actions at all. BID and RTC are the same in this respect.

Regardless, in the simulated benchmark, I used the exact same weights for both methods. The original code has this line:
```
return jnp.sum(error * weights[None, None, :], axis=-1)  # [n, b]
```
If you look at how `weights` is computed, it is zero in the padding region (same as RTC).

**Comment by wenxichen2746** [comment]
Thanks for the clarification! Yes, I meant the backward loss should be evaluated on the overlapping part. I see the exponential schedule is also truncated to zeros after attention_prefix_horizon, the BID backward loss is correctly implemented. So the performance gain for BID must originate from diverse experts' behaviors. Unfortunately, we did not get a chance to test an ablation study on this. We will post more on the diverse reward settings in Kinetix in future work. Thanks again for your time.
